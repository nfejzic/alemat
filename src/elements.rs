//! Elements found in the MathML Core Specification. These elements are called MathML Core
//! elements.

mod maction;
mod math;
mod merror;
mod mmultiscripts;
mod mphantom;
mod mrow;
mod mstyle;

pub mod grouping {
    pub use super::maction::*;
    pub use super::math::*;
    pub use super::merror::*;
    pub use super::mmultiscripts::Prescripts;
    pub use super::mphantom::*;
    pub use super::mrow::*;
    pub use super::mstyle::*;
}

mod mroot;

pub mod radicals {
    pub use super::mroot::*;
}

mod msubsup;
mod munderover;

pub mod scripted {
    pub use super::mmultiscripts::Multiscripts;
    pub use super::msubsup::*;
    pub use super::munderover::*;
}

mod annotation;
mod mfrac;
mod mi;
mod mn;
mod mo;
mod mpadded;
mod ms;
mod mspace;
mod mtable;
mod mtext;

pub use annotation::*;
pub use mfrac::*;
pub use mi::*;
pub use mn::*;
pub use mo::*;
pub use mpadded::*;
pub use ms::*;
pub use mspace::*;
pub use mtable::*;
pub use mtext::*;
