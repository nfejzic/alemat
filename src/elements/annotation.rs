use std::marker::PhantomData;

use crate::{
    attributes::Attribute,
    markers::{Init, Uninit},
    MathMl,
};

/// The content of `annotation` element, either text or MathML.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum AnnotationContent {
    Text(String),
    MathMl(MathMl),
}

impl From<String> for AnnotationContent {
    fn from(value: String) -> Self {
        Self::Text(value)
    }
}

impl<T> From<T> for AnnotationContent
where
    T: Into<MathMl>,
{
    fn from(value: T) -> Self {
        Self::MathMl(value.into())
    }
}

/// An attribute of `annotation` element. Either one of the global [`Attribute`]s, or `encode`
/// attribute.
///
/// [`Attribute`]: crate::attributes::Attribute
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum AnnotationAttr {
    Global(Attribute),
    Encoding(String),
}

/// The `annotation` (and `annotation-xml`) element.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Annotation {
    content: AnnotationContent,
    attributes: Vec<AnnotationAttr>,
}

impl Annotation {
    pub fn builder() -> AnnotationBuilder<Uninit> {
        AnnotationBuilder {
            content: None,
            ..Default::default()
        }
    }
}

impl<T> From<T> for Annotation
where
    T: Into<AnnotationContent>,
{
    fn from(value: T) -> Self {
        Self {
            content: value.into(),
            attributes: Default::default(),
        }
    }
}

#[derive(Debug, Default, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct AnnotationBuilder<T> {
    content: Option<AnnotationContent>,
    attr: Vec<AnnotationAttr>,

    _marker: PhantomData<(T,)>,
}

impl<T> AnnotationBuilder<T> {
    pub fn content(self, content: impl Into<AnnotationContent>) -> AnnotationBuilder<Init> {
        AnnotationBuilder {
            content: Some(content.into()),
            attr: self.attr,
            _marker: PhantomData,
        }
    }

    pub fn attr<A>(mut self, attr: A) -> AnnotationBuilder<T>
    where
        A: IntoIterator<Item = AnnotationAttr>,
    {
        self.attr.extend(attr);
        self
    }
}

impl AnnotationBuilder<Init> {
    pub fn build(self) -> Annotation {
        Annotation {
            content: self
                .content
                .expect("Content is guaranteed to be initialized at compile time."),
            attributes: self.attr,
        }
    }
}

/// The `semantics` element is the container element that associates annotations with a MathML
/// expression. Typically, the `semantics` element has as its first child element a MathML
/// expression to be annotated while subsequent child elements represent text annotations within an
/// `annotation` element, or more complex markup annotations within an `annotation-xml` element.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Semantics {
    /// Children of the `semantics` element. Rendering is same as `mrow`.
    children: MathMl,

    /// The `semantics` element accepts the global [`Attribute`]s.
    attr: Vec<Attribute>,
}
