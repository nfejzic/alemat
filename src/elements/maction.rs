use std::marker::PhantomData;

use crate::{
    attributes::Attribute,
    markers::{Init, Uninit},
    MathMl,
};

/// The `maction` element accepts the global [`Attribute`]s as well as `selection` and
/// `actiontype`.
///
/// [`Attribute`]: crate::attributes::Attribute
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum ActionAttr {
    Global(Attribute),
    /// The child element currently visible, only taken into account for `actiontype="toggle"` or
    /// non-standard actiontype values. The default value is `1`, which is the first child element.
    Selection(String),

    /// The action which specifies what happens for this element. Special behavior for the
    /// following values were implemented by some browsers:
    ///
    /// * `statusline`: If there is a click on the expression or the reader moves the pointer over
    ///   it, the message is sent to the browser's status line. The syntax is:
    ///
    ///   ```html
    ///   <maction actiontype="statusline"> expression message </maction>.
    ///   ```
    ///
    /// * `toggle`: When there is a click on the subexpression, the rendering alternates the
    ///    display of selected subexpressions. Therefore each click increments the selection value.
    ///    The syntax is:
    ///
    ///   ```html
    ///   <maction actiontype="toggle" selection="positive-integer">
    ///       expression1 expression2 expressionN
    ///   </maction>.
    ///   ```
    ActionType(String),
}

/// Historically, the `maction` element provides a mechanism for binding actions to expressions.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Action {
    content: MathMl,
    attributes: Vec<ActionAttr>,
}

impl Action {
    /// Create new `maction` element.
    pub fn with_mathml(math: impl Into<MathMl>) -> Self {
        Self {
            content: math.into(),
            attributes: Default::default(),
        }
    }

    pub fn builder() -> ActionBuilder<Uninit> {
        ActionBuilder::default()
    }
}

impl<T> From<T> for Action
where
    T: Into<MathMl>,
{
    fn from(value: T) -> Self {
        Self {
            content: value.into(),
            attributes: Default::default(),
        }
    }
}

#[derive(Debug, Default, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct ActionBuilder<T> {
    content: Option<MathMl>,
    attributes: Vec<ActionAttr>,

    _marker: PhantomData<(T,)>,
}

impl<T> ActionBuilder<T> {
    pub fn content(self, content: impl Into<MathMl>) -> ActionBuilder<Init> {
        ActionBuilder {
            content: Some(content.into()),
            attributes: self.attributes,
            _marker: PhantomData,
        }
    }

    pub fn attr<I, A>(mut self, attributes: I) -> Self
    where
        I: IntoIterator<Item = A>,
        A: Into<ActionAttr>,
    {
        self.attributes
            .extend(attributes.into_iter().map(Into::into));

        self
    }
}

impl ActionBuilder<Init> {
    pub fn build(self) -> Action {
        Action {
            content: self.content.expect("Content is guaranteed to be init."),
            attributes: self.attributes,
        }
    }
}
