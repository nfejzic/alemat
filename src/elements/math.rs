use std::marker::PhantomData;

use crate::{
    attributes::Attribute,
    markers::{Init, Uninit},
    DisplayAttr, MathMl,
};

/// An attribute of `math` element. Either one of the global [`Attribute`]s, `display` or `alttext`
/// attribute.
///
/// [`Attribute`]: crate::attributes::Attribute
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum MathAttr {
    Global(Attribute),
    Display(DisplayAttr),
    AltText(String),
}

// TODO: figure out if this element is actually needed.
/// MathML specifies a single top-level or root math element, which encapsulates each instance of
/// MathML markup within a document. All other MathML content must be contained in a `<math>`
/// element.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Math {
    content: MathMl,
    attributes: Vec<MathAttr>,
}

impl Math {
    /// Create new `math` element.
    pub fn with_mathml(math: impl Into<MathMl>) -> Self {
        Self {
            content: math.into(),
            attributes: Default::default(),
        }
    }

    pub fn builder() -> MathBuilder<Uninit> {
        MathBuilder::default()
    }
}

impl<T> From<T> for Math
where
    T: Into<MathMl>,
{
    fn from(value: T) -> Self {
        Self {
            content: value.into(),
            attributes: Default::default(),
        }
    }
}

#[derive(Debug, Default, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct MathBuilder<T> {
    content: Option<MathMl>,
    attributes: Vec<MathAttr>,

    _marker: PhantomData<(T,)>,
}

impl<T> MathBuilder<T> {
    pub fn content(self, content: impl Into<MathMl>) -> MathBuilder<Init> {
        MathBuilder {
            content: Some(content.into()),
            attributes: self.attributes,

            _marker: PhantomData,
        }
    }

    pub fn attr<I, A>(mut self, attr: I) -> Self
    where
        I: IntoIterator<Item = A>,
        A: Into<MathAttr>,
    {
        self.attributes.extend(attr.into_iter().map(Into::into));
        self
    }
}

impl MathBuilder<Init> {
    pub fn build(self) -> Math {
        Math {
            content: self.content.expect("Content is guaranteed to be init."),
            attributes: self.attributes,
        }
    }
}
