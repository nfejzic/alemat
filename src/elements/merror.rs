use std::marker::PhantomData;

use crate::{
    attributes::Attribute,
    markers::{Init, Uninit},
    Element, Elements,
};

use super::IntoElements;

/// The merror element displays its contents as an ”error message”. The intent of this element is
/// to provide a standard way for programs that generate MathML from other input to report syntax
/// errors in their input.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Error {
    content: Elements,
    attributes: Vec<Attribute>,
}

impl Error {
    /// Create new `merror` element.
    pub fn with_mathml(math: impl IntoElements) -> Self {
        Self {
            content: math.into_elements(),
            attributes: Default::default(),
        }
    }

    /// Create a builder for [`Error`] element.
    pub fn builder() -> ErrorBuilder<Uninit> {
        ErrorBuilder::default()
    }

    /// Get a reference to the inner content of the [`Error`] element.
    pub fn content(&self) -> &[Element] {
        &self.content
    }

    /// Get a reference to all attributes of the [`Error`] element.
    pub fn attributes(&self) -> &[Attribute] {
        &self.attributes
    }
}

impl From<Element> for Error {
    fn from(value: Element) -> Self {
        Self {
            content: Elements(vec![value]),
            attributes: Default::default(),
        }
    }
}

impl From<Elements> for Error {
    fn from(value: Elements) -> Self {
        Self {
            content: value,
            attributes: Default::default(),
        }
    }
}

crate::element_from_type!(Error => Error);

/// Builder of the [`Error`] element.
#[derive(Debug, Default, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct ErrorBuilder<T> {
    content: Option<Elements>,
    attributes: Vec<Attribute>,

    _marker: PhantomData<(T,)>,
}

impl<T> ErrorBuilder<T> {
    /// Set the content of the [`Error`] element.
    pub fn content(self, content: impl IntoElements) -> ErrorBuilder<Init> {
        ErrorBuilder {
            content: Some(content.into_elements()),
            attributes: self.attributes,

            _marker: PhantomData,
        }
    }

    /// Add attributes.
    pub fn attr<I, A>(mut self, attr: I) -> Self
    where
        I: IntoIterator<Item = A>,
        A: Into<Attribute>,
    {
        self.attributes.extend(attr.into_iter().map(Into::into));
        self
    }
}

impl ErrorBuilder<Init> {
    /// Build [`Error`] element.
    pub fn build(self) -> Error {
        Error {
            content: self.content.expect("Content is guaranteed to be init."),
            attributes: self.attributes,
        }
    }
}
