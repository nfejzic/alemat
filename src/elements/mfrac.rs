use crate::{attributes::Attribute, MathMl};

/// An attribute of `mfrac` element. Either one of the global [`Attribute`]s, or `linethickness`
/// attribute.
///
/// [`Attribute`]: crate::attributes::Attribute
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum FracAttr {
    /// Global attribute.
    Global(Attribute),

    /// The linethickness attribute indicates the fraction line thickness to use for the fraction
    /// bar.
    /// It must have a value that is a valid
    /// [<length-percentage>](https://www.w3.org/TR/css-values-4/#typedef-length-percentage).
    LineThickness(String),
}

/// The merror element displays its contents as an ”error message”. The intent of this element is
/// to provide a standard way for programs that generate MathML from other input to report syntax
/// errors in their input.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Frac {
    content: MathMl,
    attributes: Vec<FracAttr>,
}

impl Frac {
    /// Create new `mfrac` element.
    pub fn with_mathml(math: impl Into<MathMl>) -> Self {
        Self {
            content: math.into(),
            attributes: Default::default(),
        }
    }
}

impl<T> From<T> for Frac
where
    T: Into<MathMl>,
{
    fn from(value: T) -> Self {
        Self {
            content: value.into(),
            attributes: Default::default(),
        }
    }
}
