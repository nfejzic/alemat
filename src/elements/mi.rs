use crate::attributes::Attribute;

/// An attribute of `mi` element. Either one of the global [`Attribute`]s, or `mathvariant`
/// attribute.
///
/// [`Attribute`]: crate::attributes::Attribute
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum IdentAttr {
    /// Global attribute.
    Global(Attribute),

    /// The linethickness attribute indicates the fraction line thickness to use for the fraction
    /// bar.
    /// It must have a value that is a valid
    /// [<length-percentage>](https://www.w3.org/TR/css-values-4/#typedef-length-percentage).
    MathVariant,
}

/// The `mi` (`Ident`) element represents a symbolic name or arbitrary text that should be rendered
/// as an identifier. Identifiers can include variables, function names, and symbolic constants.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Ident {
    ident: String,
    attributes: Vec<IdentAttr>,
}

impl<T> From<T> for Ident
where
    T: Into<String>,
{
    fn from(value: T) -> Self {
        Self {
            ident: value.into(),
            attributes: Default::default(),
        }
    }
}
