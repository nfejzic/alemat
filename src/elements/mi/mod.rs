mod dict;

use std::marker::PhantomData;

use crate::{
    attributes::Attribute,
    markers::{Init, Uninit},
};

/// The `mi` (`Ident`) element represents a symbolic name or arbitrary text that should be rendered
/// as an identifier. Identifiers can include variables, function names, and symbolic constants.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Ident {
    ident: String,
    attributes: Vec<Attribute>,
}

impl Ident {
    /// Create a builder for [`Ident`] element.
    pub fn builder() -> IdentBuilder<Uninit> {
        IdentBuilder::default()
    }

    /// Get a reference to the inner content of the [`Ident`] element.
    pub fn ident(&self) -> &str {
        &self.ident
    }

    /// Get a reference to all attributes of the [`Ident`] element.
    pub fn attributes(&self) -> &[Attribute] {
        &self.attributes
    }
}

impl<T> From<T> for Ident
where
    T: Into<String>,
{
    fn from(value: T) -> Self {
        Self {
            ident: value.into(),
            attributes: Default::default(),
        }
    }
}

crate::element_from_type!(Ident => Ident);

/// Builder of the [`Ident`] element.
#[derive(Debug, Default, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct IdentBuilder<T> {
    ident: Option<String>,
    attributes: Vec<Attribute>,
    _marker: PhantomData<(T,)>,
}

impl<T> IdentBuilder<T> {
    /// Set the identifier string for the [`Ident`] element.
    pub fn ident(self, ident: impl Into<String>) -> IdentBuilder<Uninit> {
        IdentBuilder {
            ident: Some(ident.into()),
            attributes: self.attributes,
            _marker: PhantomData,
        }
    }

    /// Add a attributes.
    pub fn attr<I, A>(mut self, attr: I) -> Self
    where
        I: IntoIterator<Item = A>,
        A: Into<Attribute>,
    {
        self.attributes.extend(attr.into_iter().map(Into::into));
        self
    }
}

impl IdentBuilder<Init> {
    /// Build the [`Ident`] element.
    pub fn build(self) -> Ident {
        Ident {
            ident: self.ident.expect("Content is guaranteed to be init."),
            attributes: self.attributes,
        }
    }
}
