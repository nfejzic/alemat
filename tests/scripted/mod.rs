use alemat::elements::scripted::{SubSup, UnderOver};
use alemat::elements::{Ident, Num, Operator};
use alemat::{MathMl, MathMlAttr, MathMlFormatter};

#[test]
fn subsup() {
    let out = MathMl::with_content(
        SubSup::builder()
            .subscript(Ident::from("0"))
            .supscript(Ident::from("1"))
            .base(Ident::from("x"))
            .build(),
    )
    .render(&mut MathMlFormatter);

    crate::snap_test!(out, name: "subsup_subsup");
}

#[test]
fn subsup_integral() {
    let out = MathMl::with_content(alemat::children![
        SubSup::builder()
            .base(Operator::integral())
            .subscript(Ident::from("0"))
            .supscript(Ident::from("1"))
            .build(),
        Ident::from("x"),
        Ident::from("dx"),
    ])
    .with_attr([MathMlAttr::Display(String::from("block"))])
    .render(&mut MathMlFormatter);

    crate::snap_test!(out, name: "subsup_integral");
}

#[test]
fn subsup_summation() {
    let out = MathMl::with_content(alemat::children![
        SubSup::builder()
            .base(Operator::sum())
            .subscript(alemat::row![Ident::from("i"), Operator::eq(), Num::from(0)])
            .supscript(Ident::from("k"))
            .build(),
        SubSup::builder()
            .base(Num::from(2))
            .supscript(Ident::from("i"))
            .build()
    ])
    .with_attr([MathMlAttr::Display(String::from("block"))])
    .render(&mut MathMlFormatter);

    crate::snap_test!(out, name: "subsup_summation");
}

#[test]
fn underover() {
    let out = MathMl::with_content(
        UnderOver::builder()
            .under(Ident::from("0"))
            .over(Ident::from("1"))
            .expr(Ident::from("x"))
            .build(),
    )
    .render(&mut MathMlFormatter);

    crate::snap_test!(out, name: "underover_underover");
}

#[test]
fn underover_integral() {
    let out = MathMl::with_content(alemat::children![
        UnderOver::builder()
            .expr(Operator::integral())
            .under(Ident::from("0"))
            .over(Ident::from("1"))
            .build(),
        Ident::from("x"),
        Ident::from("dx"),
    ])
    .with_attr([MathMlAttr::Display(String::from("block"))])
    .render(&mut MathMlFormatter);

    crate::snap_test!(out, name: "underover_integral");
}

#[test]
fn underover_summation() {
    let out = MathMl::with_content(alemat::children![
        UnderOver::builder()
            .expr(Operator::sum())
            .under(alemat::row![Ident::from("i"), Operator::eq(), Num::from(0)])
            .over(Ident::from("k"))
            .build(),
        SubSup::builder()
            .base(Num::from(2))
            .supscript(Ident::from("i"))
            .build()
    ])
    .with_attr([MathMlAttr::Display(String::from("block"))])
    .render(&mut MathMlFormatter);

    crate::snap_test!(out, name: "underover_summation");
}
