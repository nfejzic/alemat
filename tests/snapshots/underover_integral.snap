---
source: tests/scripted/mod.rs
expression: out
---
<math display="block"><munderover ><mo>âˆ«</mo><mi>0</mi><mi>1</mi></munderover><mi>x</mi><mi>dx</mi></math>
